features/26-mfa/api.md

```diff
@@ -0,0 +1,961 @@
+  #
+  # 'required' means MFA is always required during authentication.
+  enforcement: 'optional'
+  # The total maximum number of authenticator the user can register.
+  # Valid values is [0, 15].
+  # Default is 1.
+  maximum: 1
+  totp:
+    # The maximum number of TOTP authenticator the user can register.
+    # Valid values is [0, 5].
+    # Default is 0.
+    maximum: 0
+  oob:
+    sms:
+      # The maximum number of OOB-SMS authenticator the user can register.
+      # Valid values is [0, 5].
+      # Default is 0.
+      maximum: 0
+    email:
+      # The maximum number of OOB-Email authenticator the user can register.
+      # Valid values is [0, 5].
+      # Default is 0.
+      maximum: 0
+  bearer_token:
+    # How many days the bearer token is valid. Default is 30.
+    expire_in_days: 30
+  recovery_code:
+    # How many recovery code should be generated. Default is 16.
+    count: 16
+    # Enable listing the existing recovery code.
+    # Some services like GitHub allow listing the existing recovery code.
+    # By default listing is disabled, the user must regenerate a new set of
+    # recovery code.
+    list_enabled: false
+```
+
+# Configuration Recipes
+
+## At most 1 TOTP or 1 OOB-SMS or 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP and 1 OOB-SMS and 1 OOB-Email
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  maximum: 3
+  totp:
+    maximum: 1
+  oob:
+    sms:
+      maximum: 1
+    email:
+      maximum: 1
+```
+
+## At most 1 TOTP
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  totp:
+    maximum: 1
+```
+
+## At most 1 OOB-SMS
+
+```yaml
+mfa:
+  enabled: true
+  enforcement: "optional"
+  oob:
+    sms:
+      maximum: 1
+```
+
+# Endpoints
+
+## POST /mfa/totp/new
+
+### Request body
+
+```JSON
+{
+  "display_name": "my TOTP"
+}
+```
+
+- `display_name`: The display name of the new TOTP authenticator.
+- `authn_session_token`: Required only when there is no access token.
+
+### Response body
+
+```JSON
+{
+  "authenticator_id": "xxx",
+  "authenticator_type": "totp",
+  "secret": "secret"
+}
+```
+
+- `secret`: The secret encoded in Base32 without padding.
+
+### Description
+
+Delete all existing inactive TOTP authenticators and create a new TOTP authenticator.
+
+## POST /mfa/totp/activate
+
+### Request body
+
+```JSON
+{
+  "otp": "123456"
+}
+```
+
+- `otp`: The one-time password generated by the secret of the authenticator.
+- `authn_session_token`: Required only when there is no access token.
+
+### Response body
+
+```JSON
+{
+  "recovery_codes": ["xxx", "xxx"]
+}
+```
+
+- `recovery_codes`: Present only when the activated authenticator is the first active authenticator.
+
+### Description
+
+Activate the just created inactive TOTP authenticator.
+
+## GET /mfa/totp/qrcode
+
+## Query parameters
+
+- `otpauth_uri`: The otpauth URI
```

Comment:
usability-wise, would it be better if we let people do either `otpath_uri`, or:

`label`: Lable of the account
`secret`: secret (or `authenticator_id`?)
`issuer`: provider or the service this account is associated with

Reply:
We have [this function](https://skygeario.github.io/skygear-SDK-JS/docs/web.generateotpauthuri) in the client SDK, which effectively turn [label, secret and issuer](https://skygeario.github.io/skygear-SDK-JS/docs/web.generateotpauthurioptions) into `otpauth_uri`.

`authenticator_id` should **NOT** be used because it should be impossible to reveal the secret after the creation. The secret is only revealed once during creation.